---
title: "Introduction to the Isosceles package"
author:
- name: Michal Kabza
  affiliation:
    - Roche Informatics, F. Hoffmann-La Roche Ltd., Poznań, Poland
  email: michal.kabza@contractors.roche.com
- name: Tim Sterne-Weiler
  affiliation:
    - Department of Oncology Bioinformatics, Genentech Inc., South San Francisco, USA
    - Department of Discovery Oncology, Genentech Inc., South San Francisco, USA
  email: sternewt@gene.com
output:
  BiocStyle::html_document:
    toc_depth: 1
    toc_float: true
package: Isosceles
vignette: >
  %\VignetteIndexEntry{Introduction to the Isosceles package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "##"
)
```

# Introduction 

Isosceles (**Iso**form **S**ingle-**Ce**ll and **L**ong-read **E**xpression 
**S**uite) is an R package dedicated to transcript detection and quantification 
from long read RNA-seq data, supporting both bulk and single-cell technologies.

Isosceles can be installed using the following commands:
```r
install.packages(c("BiocManager", "devtools"))
BiocManager::install(c("scran", "scater", "uwot", "dittoSeq", "DEXSeq", 
                       "Nebulosa", "ggbio", "BiocStyle"))
devtools::install_github("timbitz/Isosceles", dependencies = TRUE,
                         upgrade = TRUE)
```

Load the Isosceles package:
```{r, message = FALSE}
library(Isosceles)
```

# Input data

Isosceles requires the following input files:

  * BAM file(s) containing aligned long reads
    * We recommend [minimap2](https://github.com/lh3/minimap2) for all long-read alignments.
    * As Isosceles doesn't perform post-alignment splice junction correction, we also recommend running minimap2 with the '--junc-bed' flag. The intron position BED file required by it can be easily created using the `gtf_to_intron_bed` function.
    * Isosceles requires the BAM file for single-cell data to contain a tag storing the cell barcode. Several pipelines are suitable for this purpose, including [Sicelore](https://github.com/ucagenomix/sicelore), which is the default tag name (`BC` tag), but it can be changed using the *barcode_tag* argument of the `bam_to_tcc` function.
  * FASTA file containing the genomic sequences
  * GTF file containing the genomic annotations
    * Isosceles requires the chromosome names to match between the FASTA file and the GTF file. No automatic conversion between alternative chromosome naming conventions (e.g. '1' and 'chr1') is performed.

# Run modes

Isosceles can used in of the following run modes, specified when the `bam_to_tcc` function is run:

  * **strict** - only annotated reference transcripts will be quantified
  * **de_novo_strict** - de novo transcripts enabled, but all splice sites must be known (i.e. found in reference annotations or provided by the user).
  * **de_novo_loose** - de novo transcripts enabled, but all splice sites must be known or reproducibly passing filters in the aligned reads

# Genome annotations

Isosceles selects the appropriate set of transcripts to quantify, depending on the given run mode, using the genomic annotations that are prepared with the `prepare_transcripts` function. This function needs a GTF file with reference annotations, and, optionally, transcript structures extracted from BAM files using the `bam_to_read_structures` function (which is necessary for detecting de novo transcripts). If de novo transcript detection is enabled, the numbers of spliced reads assigned to different splicing compatibility levels (described in detail in the Isosceles paper) can be plotted using the `plot_splicing_support_levels` function.

There are several important aspects of the `prepare_transcripts` function to consider:

  * Reference annotations from Ensembl or GENCODE are generally recommended. However, Isosceles can also work with merged GTF files that combine both reference and de novo detected transcripts from an external source. Examples of programs that produce these merged files include [StringTie](https://ccb.jhu.edu/software/stringtie/) and [IsoQaunt](https://github.com/ablab/IsoQuant). Utilizing merged annotations as reference can improve the sensitivity of discovering new transcripts without significantly increasing the false discovery rate. More details about the benchmarks can be found in the Isosceles paper and the [Isosceles_paper](https://github.com/timbitz/Isosceles_paper) repository.
  * Isosceles assigns new IDs to all transcripts (whether they are reference or de novo detected) based on stable hash identifiers. This simplifies the process of matching de novo transcripts across different data sets from the same genome build, irrespective of other experimental variables. The corresponding reference IDs compatible with each transcript can be found in transcript level SummarizedExperiment object's rowData (`rowData(se)$compatible_tx`).
  * Genes and transcripts are merged according to specific rules:
    * If annotated genes share introns, they are merged and assigned a new ID and symbol, comprising the original IDs and symbols separated by commas.
    * Annotated transcripts with identical intron structure and identical transcript start and end bins (default: 50 bp) are considered identical by Isosceles and also merged, and given a unique transcript ID. 

# Isosceles transcript IDs

The Isosceles transcript ID (e.g. 'ISOT-2a9c-c3db-71b4-c3f2:s100628650:e100636850:AP:FL') consists of several parts, separated by colons:

  * Stable hash identifier, starting with the constant 'ISOT' prefix. For spliced transcripts, the identifier contains the first 16 characters of the MD5 hash of transcript's comma-separated intron positions, while for the unspliced transcripts it consists of 12 zeroes and the first 4 characters of the MD5 hash of transcript's chromosome and strand. For convenience, the hash identifier is separated into 4 character segments by dashes.
  * The transcript start bin position, prefixed by the 's' character. The default bin size is 50 bp, but it can be changed using the *bin_size* argument of the `prepare_transcripts` function. The start of the bin is always used, ensuring that it's smaller or equal to the transcript start position.
  * The transcript end bin position, prefixed by the 'e' character. The default bin size is 50 bp, but it can be changed using the *bin_size* argument of the `prepare_transcripts` function. Unlike the start positions, the end of the bin is always used, ensuring that it's bigger or equal to the transcript end position.
  * Splicing compatibility code of the transcript. The full list of these codes can be found in the Isosceles paper, but the following ones are actually used in at least one run mode:
    * AP (Annotated Paths) - reference transcripts
    * EC (Edge Compatible) - de novo transcripts where all introns are known (i.e. found in reference annotations or provided by the user)
    * NC (Node Compatible) - de novo transcripts where all splice sites are known
    * DN (De-novo Node) - de novo transcripts where all splice sites are known or reproducibly passing filters in the aligned reads
  * Truncation status code of the transcript. The full list of these codes can be found in the Isosceles paper, but only the FL (Full-Length) transcripts are used by Isosceles for quantification.

The comma separated list of compatible reference transcript IDs can be found in transcript level SummarizedExperiment object's rowData (`rowData(se)$compatible_tx`).

# Output

The main output of Isosceles consists of SummarizedExperiment objects, created at various resolution levels:

  * **TCC (Transcript Compatibility Counts)** - This is an intermediate quantification stage of the data, formed directly from the BAM files with the `bam_to_tcc` function. The values here are used to derive transcript quantification estimates via the EM algorithm. The `pseudobulk_tcc` function (for summarizing data at the pseudo-bulk level), the `pseudotime_tcc` function (for summarizing data at the pseudo-time window level) and the `neighborhood_tcc` function (for combining counts from neighboring cells) utilize TCC SummarizedExperiment objects as both input and output.
  * **transcript** - At the transcript level, expression values are computed using the `tcc_to_transcript` function, applying the EM (Expectation–Maximization) algorithm. By default, normalization is done using effective transcript lengths, but it can be turned off for UMI-based protocol data (commonly in scRNA-Seq datasets) by setting the *use_length_normalization* argument to FALSE. You can export transcript annotations from the SummarizedExperiment object to a GTF file using the `export_gtf` function.
  * **gene** - Gene-level expression values are derived using the `tcc_to_gene` function. This process is typically very swift as it doesn't necessitate running the EM algorithm.
  * **PSI (Percent Spliced In)** - PSI values are computed for unique transcriptomic regions of each gene (such as transcription start sites or core exons) using the `transcript_to_psi` function. This requires a transcript level SummarizedExperiment object as input.

Each of these levels provides specific insights and functionalities within the Isosceles framework, catering to different analytical requirements.

Data in the SummarizedExperiment objects can be accessed using the following commands:

* **assayNames(se)** - assay names available at the given resolution level
* **assay(se, "counts")** - raw count data matrix
* **assay(se, "tpm")** - TPM (Transcripts Per Million) normalized expression value matrix
* **assay(se, "relative_expression")** - relative expression value matrix. All the values are in the 0 to 1 range, and represent the ratios of the TPM values of the analyzed features (e.g. transcripts) and the TPM values of genes they belong to
* **assay(se, "psi")** - PSI (Percent Spliced In) value matrix
* **rowData(se)** - additional data regarding the analyzed features (e.g. reference transcript IDs for transcript level data)
* **rowRanges(se)** - a GRangesList object (transcript level data) or a GRanges object (PSI level data) containing the genomic positions of the analyzed features

# Parallelization

Many functions in the Isosceles package can make use of multiple CPU cores, specified using the *ncpu* argument. The following tips should help you find the optimal *ncpu* value for your analysis:

* Functions that take BAM files as an input (`bam_to_read_structures` and `bam_to_tcc`) can process each of them on a separate CPU core.
* The `tcc_to_transcript` function parallelizes the EM algorithm computations at the SummarizedExperiment column level (samples for bulk RNA-Seq, cells for scRNA-Seq data).
* In case of the `transcript_to_psi` function, PSI region detection can be parallelized at the gene level.

# Bulk RNA-Seq analysis

## Preparing input data

Get the input file paths for a small bulk RNA-Seq dataset: 
```{r, message = FALSE, warning = FALSE}
bam_file <- system.file(
  "extdata", "bulk_rnaseq.bam",
  package = "Isosceles"
)
gtf_file <- system.file(
  "extdata", "bulk_rnaseq.gtf",
  package = "Isosceles"
)
genome_fasta_file <- system.file(
  "extdata", "bulk_rnaseq.fa",
  package = "Isosceles"
)
```

## Basic Isosceles analysis

Extract read data from the BAM files:
```{r, message = FALSE, warning = FALSE}
bam_files <- c(Sample = bam_file)
bam_parsed <- bam_to_read_structures(
    bam_files = bam_files
)
```

Prepare transcript data for the analysis:
```{r, message = FALSE, warning = FALSE}
transcript_data <- prepare_transcripts(
    gtf_file = gtf_file,
    genome_fasta_file = genome_fasta_file,
    bam_parsed = bam_parsed,
    min_bam_splice_read_count = 2,
    min_bam_splice_fraction = 0.01
)
```

Create a TCC (Transcript Compatibility Counts) SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_tcc <- bam_to_tcc(
    bam_files = bam_files,
    transcript_data = transcript_data,
    run_mode = "de_novo_loose",
    min_read_count = 1,
    min_relative_expression = 0
)
se_tcc
```

Create a transcript-level SummarizedExperiment object using the EM algorithm:
```{r, message = FALSE, warning = FALSE}
se_transcript <- tcc_to_transcript(
    se_tcc = se_tcc,
    use_length_normalization = TRUE
)
se_transcript
```

Create a gene-level SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_gene <- tcc_to_gene(
    se_tcc = se_tcc
)
se_gene
```

# scRNA-Seq analysis

## Preparing input data

Get the input file paths for a small scRNA-Seq dataset: 
```{r, message = FALSE, warning = FALSE}
bam_file <- system.file(
    "extdata", "scrnaseq.bam",
    package = "Isosceles"
)
gtf_file <- system.file(
    "extdata", "scrnaseq.gtf.gz",
    package = "Isosceles"
)
genome_fasta_file <- system.file(
    "extdata", "scrnaseq.fa.gz",
    package = "Isosceles"
)
```

## Basic Isosceles analysis

Extract read data from the BAM files:
```{r, message = FALSE, warning = FALSE}
bam_files <- c(Sample = bam_file)
bam_parsed <- bam_to_read_structures(
    bam_files = bam_files
)
```

Prepare transcript data for the analysis:
```{r, message = FALSE, warning = FALSE}
transcript_data <- prepare_transcripts(
    gtf_file = gtf_file,
    genome_fasta_file = genome_fasta_file,
    bam_parsed = bam_parsed,
    min_bam_splice_read_count = 1,
    min_bam_splice_fraction = 0.01
)
```

Create a TCC (Transcript Compatibility Counts) SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_tcc <- bam_to_tcc(
    bam_files = bam_files,
    transcript_data = transcript_data,
    run_mode = "de_novo_loose",
    min_read_count = 1,
    min_relative_expression = 0,
    is_single_cell = TRUE,
    barcode_tag = "BC"
)
se_tcc
```

Create a transcript-level SummarizedExperiment object using the EM algorithm:
```{r, message = FALSE, warning = FALSE}
se_transcript <- tcc_to_transcript(
    se_tcc = se_tcc,
    use_length_normalization = FALSE
)
se_transcript
```

Create a gene-level SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_gene <- tcc_to_gene(
    se_tcc = se_tcc
)
se_gene
```

## Pseudo-bulk analysis

Create a pseudobulk TCC SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
# Randomly choose one of two labels ('1' or '2') for each cell
set.seed(42)
cell_labels <- sample(1:2, ncol(se_tcc), replace = TRUE)
# Prepare pseudobulk TCC data for given labels
se_pseudobulk_tcc <- pseudobulk_tcc(
    se_tcc = se_tcc,
    cell_labels = cell_labels
)
se_pseudobulk_tcc
```

Create a pseudobulk transcript-level SummarizedExperiment object using the EM algorithm:
```{r, message = FALSE, warning = FALSE}
se_pseudobulk_transcript <- tcc_to_transcript(
    se_tcc = se_pseudobulk_tcc,
    use_length_normalization = FALSE
)
se_pseudobulk_transcript
```

Create a pseudobulk gene-level SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_pseudobulk_gene <- tcc_to_gene(
    se_tcc = se_pseudobulk_tcc
)
se_pseudobulk_gene
```

## Merging neighboring cell data

Isosceles also allows adding counts from neighboring cells to each cell, 
potentially improving the quantification results:
```{r, message = FALSE, warning = FALSE}
# Get PCA coordinates of each cell from transcript expression data
sce_transcript <- methods::as(se_transcript, "SingleCellExperiment")
sce_transcript <- scuttle::computeLibraryFactors(sce_transcript)
sce_transcript <- scuttle::logNormCounts(sce_transcript)
set.seed(42)
sce_transcript <- scater::runPCA(sce_transcript, ncomponents = 2)
pca_mat <- SingleCellExperiment::reducedDim(sce_transcript, "PCA")
# Add TCC values from two nearest cells to each cell
se_neighbor_tcc <- neighborhood_tcc(
    se_tcc = se_tcc,
    pca_mat = pca_mat,
    k = 2
)
se_neighbor_tcc
```

Recalculate transcript expression values from merged data using the EM algorithm:
```{r, message = FALSE, warning = FALSE}
se_neighbor_transcript <- tcc_to_transcript(
    se_tcc = se_neighbor_tcc,
    use_length_normalization = FALSE
)
se_neighbor_transcript
```

# Downstream analysis

## Preparing input data

Load a transcript-level SummarizedExperiment object containing expression levels 
of 100 highly variable transcripts across 200 cells from a mix of ovarian cancer 
cell lines:
```{r, message = FALSE, warning = FALSE}
se <- readRDS(system.file("extdata", "se_transcript_100_hvts.rds",
                          package = "Isosceles"))
se
```

## Basic scRNA-Seq analysis

Run a basic scRNA-Seq data analysis using Bioconductor packages:
```{r, message = FALSE, warning = FALSE}
sce <- as(se, "SingleCellExperiment")
sce <- scuttle::computeLibraryFactors(sce)
sce <- scuttle::logNormCounts(sce)
dec <- scran::modelGeneVar(sce)
top_hvgs <- scran::getTopHVGs(dec, n = 100)
set.seed(42)
sce <- scran::denoisePCA(sce, technical = dec, subset.row = top_hvgs)
set.seed(42)
sce <- scater::runUMAP(sce, dimred = "PCA")
sce
```

Cell clustering using k-means:
```{r, message = FALSE, warning = FALSE}
set.seed(42)
colLabels(sce) <- bluster::clusterRows(reducedDim(sce, "PCA"),
                                       bluster::KmeansParam(3))
dittoSeq::dittoDimPlot(sce, "label", reduction.use = "UMAP",
                       size = 2, main = "",
                       do.label = TRUE, labels.highlight = TRUE,
                       legend.show = FALSE)
```

## Identifying marker transcripts

Identify marker transcripts for individual clusters:
```{r, message = FALSE, warning = FALSE}
markers <- scran::findMarkers(sce, groups = sce$label, test.type = "t",
                              pval.type = "all", direction = "up")
markers
head(markers[[1]])
```

Get top 3 marker transcripts for each cluster:
```{r, message = FALSE, warning = FALSE}
top_markers <- markers %>%
  lapply(function(x) rownames(x)[1:3]) %>%
  unlist() %>%
  unname()
top_markers
```

Top marker transcript heatmap:
```{r, message = FALSE, warning = FALSE}
dittoSeq::dittoHeatmap(sce, top_markers, annot.by = "label",
                       cluster_rows = TRUE, fontsize_row = 6)
```

## Transcript expression UMAP plots {.tabset}

```{r, message = FALSE, warning = FALSE, results = 'asis'}
for (transcript_id in top_markers) {
    cat("\n\n### ", transcript_id, "\n")
    umap_plot <- dittoSeq::dittoDimPlot(
        sce, transcript_id, size = 2,
        min.color = "#0072B2", max.color = "#F0E442",
        order = "increasing"
    )
    print(umap_plot)
}
```

### {-}

# PSI analysis

## Preparing input data

Load a transcript-level SingleCellExperiment object containing expression levels 
of transcripts of the CAV1 gene across 2,060 cells from a mix of ovarian cancer 
cell lines:
```{r, message = FALSE, warning = FALSE}
sce_transcript <- readRDS(system.file("extdata", "sce_transcript_cav1.rds",
                                      package = "Isosceles"))
sce_transcript
```

## PSI analysis using Isosceles

Create a PSI (Percent Spliced In) SummarizedExperiment object:
```{r, message = FALSE, warning = FALSE}
se_psi <- transcript_to_psi(sce_transcript)
se_psi
```

PSI values are calculated for the following types of regions:

  * **TSS** - transcription start sites
  * **TES** - transcription end sites
  * **CE** - core exonic regions
  * **RI** - retained intronic regions
  * **A5** - 5' alternative exonic regions
  * **A3** - 3' alternative exonic regions

## PSI region visualization

PSI value heatmap for the CAV1 gene:
```{r, message = FALSE, warning = FALSE}
region_colors <- c(
    TSS = "#FF83FF", CE = "#FF9289", RI = "#82B7FF",
    A5 = "#00D65C", A3 = "#00DAE0", TES = "#D3BA00"
)
plot_psi_heatmap(se_psi, gene_id = "ENSG00000105974",
                 region_colors = region_colors)
```

Plot the CAV1 transcript structures:
```{r, message = FALSE, warning = FALSE}
plot_psi_regions(
    se_psi = se_psi,
    se_transcript = sce_transcript,
    gene_id = "ENSG00000105974",
    region_colors = region_colors
)
```

Plot the CAV1 transcript structures with introns shrinked to max 1000 bp:
```{r, message = FALSE, warning = FALSE}
plot_psi_regions(
    se_psi = se_psi,
    se_transcript = sce_transcript,
    gene_id = "ENSG00000105974",
    max_intron_length = 1000,
    region_colors = region_colors
)
```

## PSI UMAP plots {.tabset}

```{r, message = FALSE, warning = FALSE, results = 'asis'}
# Calculate UMAP coordinates based on PSI values
sce_psi <- as(se_psi, "SingleCellExperiment")
reducedDim(sce_psi, "UMAP") <- reducedDim(sce_transcript, "UMAP")
sce_psi <- sce_psi[
    , colSums(assay(sce_psi, "psi")) > 0
]
# Create UMAP plots colored by PSI values for each gene region
for (region_id in rownames(sce_psi)) {
    cat("\n\n### ", region_id, "\n")
    umap_plot <- dittoSeq::dittoDimPlot(
        sce_psi, region_id, size = 1,
        min.color = "#0072B2", max.color = "#F0E442",
        order = "increasing"
    )
    print(umap_plot)
}
```

### {-}

# Session Info

```{r}
sessionInfo()
```
